ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsRange.scala

```diff
@@ -40,25 +49,47 @@ object EventsRange {
     if (startExclusive == Offset.beforeBegin) {
       EmptyEventSeqIdRange.startExclusive
     } else {
-      import com.daml.platform.store.Conversions.OffsetToStatement
-      // This query could be: "select min(event_sequential_id) - 1 from participant_events where event_offset > ${range.startExclusive}"
-      // however there are cases when postgres decides not to use the index. We are forcing the index usage specifying `order by event_offset`
-      SQL"select min(event_sequential_id) from participant_events where event_offset > ${startExclusive} group by event_offset order by event_offset asc limit 1"
-        .as(get[Long](1).singleOpt)(connection)
-        .map(_ - 1L)
-        .getOrElse(EmptyEventSeqIdRange.startExclusive)
+      // TODO(Leo): fuse these two queries
+      readMaxRowId(startExclusive)(connection).getOrElse(
+        readNextRowId(startExclusive)(connection)
+          .map(_ - 1L)
+          .getOrElse(
+            if (isLedgerEmpty(connection))
+              EmptyEventSeqIdRange.startExclusive
+            else
+              throw new IllegalStateException(
+                s"readLowerBound($startExclusive) returned None on non-empty ledger")
+          )
+      )
     }
 
+  private def readMaxRowId(offset: Offset)(connection: java.sql.Connection): Option[Long] = {
+    SQL"select max(event_sequential_id) from participant_events where event_offset = ${offset}"
+      .as(get[Option[Long]](1).single)(connection)
+  }
+
+  private def readNextRowId(offset: Offset)(connection: java.sql.Connection): Option[Long] = {
+    SQL"select min(event_sequential_id) from participant_events where event_offset > ${offset}"
```

Comment:
yes, that is true... on postgres 12 we had to use `order by event_offset` to force it to use the index. I was planning on returning this but it would require a bit more testing, did not want to block the release.